docs/pages/2020_Intro_Week/sections/workflows_adv.rst

```diff
@@ -1,6 +1,339 @@
+Since there is not one output port that is shared by all process classes, it is less critical to use the ``namespace`` argument when exposing outputs.
+
+That takes care of exposing the port specification of the wrapped process class in a very efficient way.
+To efficiently retrieve the inputs that have been passed to the process, one can use the :meth:`~aiida.engine.processes.process.Process.exposed_inputs` method.
+Note the past tense of the method name.
+The method takes a process class and an optional namespace as arguments, and will return the inputs that have been passed into that namespace when it was launched.
+This utility now allows us to simplify the ``setup`` outline step that we have shown before:
+
+.. code-block:: python
+
+    def setup(self):
+        """Call the `setup` of the `BaseRestartWorkChain` and then create the inputs dictionary in `self.ctx.inputs`.
+
+        This `self.ctx.inputs` dictionary will be used by the `BaseRestartWorkChain` to submit the process in the
+        internal loop.
+        """
+        super().setup()
+        self.ctx.inputs = self.exposed_inputs(ArithmeticAddCalculation, 'add')
+
+This way we don't have to manually fish out all the individual inputs from the ``self.inputs`` but have to just call this single method, saving a lot of time and lines of code.
+
+
+Error handling
+==============
+So far you have seen how easy it is to get a work chain up and running that will run a subprocess using the ``BaseRestartWorkChain``.
+However, the whole point of this exercise, as described in the introduction, was for the work chain to be able to deal with *failing* processes, yet in the previous example it finished without any problems.
+What would have happened if the subprocess had failed?
+If the computed sum of the inputs ``x`` and ``y`` is negative, the ``ArithmeticAddCalculation`` fails with exit code ``410`` which corresponds to ``ERROR_NEGATIVE_NUMBER``.
+Let's launch the work chain with inputs that will cause the calculation to fail, e.g. by making on of the operands negative, and see what happens:
+
+.. code-block:: python
+
+    submit(ArithmeticAddBaseWorkChain, x=Int(3), y=Int(-4), code=load_code('add@localhost'))
+
+This time we will see that the work chain takes quite a different path:
+
+.. code-block:: console
+
+    ArithmeticAddBaseWorkChain<1930> Finished [402] [1:while_(should_run_process)(1:inspect_process)]
+        ├── ArithmeticAddCalculation<1931> Finished [410]
+        └── ArithmeticAddCalculation<1934> Finished [410]
+
+As expected, the ``ArithmeticAddCalculation`` failed this time with a ``410``.
+The work chain noticed the failure when inspecting the result of the subprocess in ``inspect_process``, and in keeping with its name and design, restarted the calculation.
+However, since the inputs were not changed, the calculation inevitably and wholly unexpectedly failed once more with the exact same error code.
```

Comment:
```suggestion
However, since the inputs were not changed, the calculation inevitably and wholly expectedly failed once more with the exact same error code.
```